transmart-core-db/grails-app/services/org/transmartproject/db/clinical/AggregateDataOptimisationsService.groovy

```diff
@@ -0,0 +1,263 @@
+     * @return
+     */
+    @Memoized
+    boolean isCountsPerStudyAndConceptForPatientSetEnabled() {
+        def session = (StatelessSessionImpl)sessionFactory.openStatelessSession()
+        try {
+            def connection = session.connection()
+
+            log.info "Check if view i2b2demodata.study_concept_sets exists ..."
+            def resultSet = connection.metaData.getTables(null, 'i2b2demodata', 'study_concept_sets', null)
+            if (resultSet.next()) {
+                log.info "Check if view i2b2demodata.patient_num_bounds exists ..."
+                def rs = connection.metaData.getTables(null, 'i2b2demodata', 'patient_num_bounds', null)
+                def result = rs.next()
+                if (!result) {
+                    log.info "View i2b2demodata.patient_num_bounds not found."
+                }
+                return result
+            }
+            log.info "View i2b2demodata.study_concept_sets not found."
+            return false
+        } finally {
+            session.close()
+        }
+    }
+
+    String bitcount_function(String arguments) {
+        if (bitCountFunctionAvailable) {
+            "public.pg_bitcount(${arguments})"
+        } else {
+            "length(replace((${arguments})::text, '0', ''))"
+        }
+    }
+
+    /**
+     * Implementation using bit sets. Only returns patient counts, not observation counts.
+     * @param constraint
+     * @param user
+     * @return
+     */
+    @CompileStatic
+    Map<String, Map<String, Counts>> countsPerStudyAndConceptForPatientSet(PatientSetConstraint constraint, User user) {
+        def resultInstanceId = constraint.patientSetId
+        def patientSet = multidimensionalDataResourceService.findQueryResult(resultInstanceId, user)
+        long size = patientSet.setSize
+        log.info "Counts per study and concept for patient set ${resultInstanceId} (${size} subjects) using bitsets ..."
+
+        Set<String> studyIds = studiesResource.getStudies(user).stream()
+                .map({MDStudy study -> study.name})
+                .collect(Collectors.toSet())
+
+        def t1 = new Date()
+        def session = (StatelessSessionImpl)sessionFactory.openStatelessSession()
+        try {
+            session.connection().autoCommit = false
+
+            // Drop temporary table if it exists
+            session.connection().prepareStatement('drop table if exists patient_set_bitset;').execute()
+
+            def bounds = patientNumBounds // minimum patient num needs to be the same as the one used for creating the view
+            log.info "Create temporary table for patient set ${resultInstanceId}. Patient nums: min=${bounds.min}, max=${bounds.max}"
+
+            int number_of_bits = bounds.max + 1 - bounds.min
+            def tempTableStatement = session.connection().prepareStatement(
+                    """create temporary table patient_set_bitset as
```

Comment:
What about having `patient_set_bitset` as regular table and do this calculation on patient set creation rather than here.

Reply:
Covered by [TMT-157](https://jira.thehyve.nl/browse/TMT-157)